Update Subscribers with #[AsDoctrineListener]

DennisdeBest opened the pull request:
To remove the deprecation from implementing the EventSubscriberInterface this PR is to use the #[AsDoctrineListener] attribute instead.

To do so I set the `doctrine/doctrine-bundle` to `^2.7.2`.

Running the cs-fixer, phpstan and rector showed a couple of errors so I updated Rector and implemented some of the changes that those tools required.

There were errors in the github workflows so I updated the versions of the actions accordingly.

If there is more to update let me know I would be happy to help.

--- 

Looking at the current PRs this seems to be a combination of 

https://github.com/KnpLabs/DoctrineBehaviors/pull/739

https://github.com/KnpLabs/DoctrineBehaviors/pull/741

https://github.com/KnpLabs/DoctrineBehaviors/pull/738

https://github.com/KnpLabs/DoctrineBehaviors/pull/736

DennisdeBest committed:
WIP

DennisdeBest committed:
Update Subscribers, Update Rector

DennisdeBest committed:
Set to minimum PHP 8.1 Add Symfony 7 support

DennisdeBest commented:
I updated the tests some more and added support for Symfony 7.
I dropped support for Symfony5 and PHP 8.0

This kind of takes the PR [#750](https://github.com/KnpLabs/DoctrineBehaviors/pull/750) into account aswell.

jgrygierek commented:
When it will be released?

Enz000 commented:
Hello 👋 

Any updates about this PR?  😀

DennisdeBest commented:
> Hello 👋 
> 
> Any updates about this PR?  😀

Nothing so far 😕I use my fork on multiple projects waiting for it to get merged

jzecca commented:
Thanks a lot for the PR!
Would be great to have this released.

jgrygierek commented:
I really don't understand... What is the problem with releasing this one very important feature? It is just a few minutes...

virtualize commented:
@jgrygierek The problem is, that this library does not seem to have any active maintainers for quite some time. See #711

alexpozzi commented on .github/workflows/code_checks.yaml:
```diff
@@ -31,10 +31,10 @@ jobs:
+            -   uses: shivammathur/setup-php@v2
                 with:
-                    php-version: 8.0
+                    php-version: 8.1
```
This should be reverted as Symfony 6.0 still supports PHP 8.0.
If you want to add a test on PHP 8.1 you should add it instead of just change it to it IMHO.

connorhu replied:
symfony 6.0 is "Unmaintained" https://symfony.com/releases/6.0
symfony 6.4 is maintained ATM and: "Requires: PHP 8.1.0 or higher"

alexpozzi commented on .github/workflows/linter.yaml:
```diff
@@ -11,11 +11,11 @@ jobs:
+            -   uses: shivammathur/setup-php@v2
                 with:
-                    php-version: 8.0
+                    php-version: 8.1
```
This should be reverted as Symfony 6.0 still supports PHP 8.0.
If you want to add a test on PHP 8.1 you should add it instead of just change it to it IMHO.

alexpozzi commented on .github/workflows/tests.yaml:
```diff
@@ -7,16 +7,16 @@ jobs:
-                php:
-                    - 8.0
+                operating-system: [ ubuntu-latest ]
+                php-versions: [ '8.1', '8.3' ]
```
We should still test php 8.0 as well for the same reasons.

alexpozzi commented on composer.json:
```diff
@@ -11,18 +11,18 @@
     ],
     "require": {
-        "php": ">=8.0",
+        "php": ">=8.1",
```
This should be reverted as Symfony 6 still supports PHP 8.0 unless there is a reason for it.

alexpozzi commented on composer.json:
```diff
@@ -11,18 +11,18 @@
-        "symfony/security-core": "^5.4|^6.0",
-        "symfony/framework-bundle": "^5.4|^6.0",
-        "symfony/string": "^5.4|^6.0",
+        "doctrine/doctrine-bundle": "^2.7.2",
```
Does this constraint need to be updated?

endelwar replied:
The #AsDoctrineListener attribute was introduced in DoctrineBundle 2.7.2

alexpozzi commented on src/EventSubscriber/BlameableEventSubscriber.php:
```diff
@@ -102,20 +108,20 @@ public function preUpdate(LifecycleEventArgs $lifecycleEventArgs): void
 
-        $oldValue = $entity->getUpdatedBy();
-        $entity->setUpdatedBy($user);
+        $updatedBy = $object->getUpdatedBy();
```
```suggestion
        $oldUpdatedBy = $object->getUpdatedBy();
```

alexpozzi commented on src/EventSubscriber/BlameableEventSubscriber.php:
I would keep `$entity` instead of changing it to `$object` as it's more specific to this case.

alexpozzi commented on src/EventSubscriber/TranslatableEventSubscriber.php:
I would keep `$entity` instead of changing it to `$object` as it's more specific to this case.

alexpozzi commented on src/EventSubscriber/UuidableEventSubscriber.php:
I would keep `$entity` instead of changing it to `$object` as it's more specific to this case.

alexpozzi commented on tests/ORM/TreeNodeTest.php:
Why changing variable names?

DennisdeBest replied:
this is done by Rector automatically : RenameVariableToMatchMethodCallReturnTypeRector

alexpozzi commented on tests/config/config_test.php:
I'm afraid that this change is not compatible to both Symfony 6 and Symfony 7.

alexpozzi reviewed (commented):
Thanks for your contribution!

I'm not super comfortable with this bundle as I don't use it myself, I mainly left you compatibility and maintainability comments.

In order to avoid conflicts, it would be better to avoid to change variables names (unless it has a real benefit in doing so).

endelwar commented on composer.json:
```diff
@@ -11,18 +11,18 @@
-        "symfony/framework-bundle": "^5.4|^6.0",
-        "symfony/string": "^5.4|^6.0",
+        "doctrine/doctrine-bundle": "^2.7.2",
+        "symfony/cache": "^6.0|^7.0",
+        "symfony/dependency-injection": "^6.0|^7.0",
+        "symfony/http-kernel": "^6.0|^7.0",
+        "symfony/security-bundle": "^6.0|^7.0",
+        "symfony/framework-bundle": "^6.0|^7.0",
+        "symfony/string": "^6.0|^7.0",
```
Why are versions 6.0-6.3, which are EOL, supported, while LTS version 5.4, which has support until November 2024 and security support until November 2025, is not?

alexpozzi replied:
It can be decided to introduce a breaking change in the library (like this one) and create a new major version while still supporting 5.4 (patches only) in the previous major version.
The fact that we bump the requirements here does not mean that the library stops supporting Symfony 5.4.

Supporting 6.0-6.3 makes sense as well unless there are some constraint that are creating us issues IMHO.
When writing libraries it's better to give a wider support of versions if there are no real reasons not to do it IMHO, not everyone might have upgraded yet to the latest 6.x version.

DennisdeBest commented:
@alexpozzi

Thanks for the reply.

I will take another look at it when I get back from vacation in two weeks.

jgrygierek commented:
Any news?

ernestoliberio commented:
the package is dead?

azribilel commented:
Hello @acrobat, Any news ?

DennisdeBest committed:
Set to Symfony 6.1 minimum

DennisdeBest commented:
There are some changes with Symfony 6.1 which make it difficult to use 6.0 as the lowest version. I updated the composer.json to only allow symfony 6.1 and php 8.1.

The changes  of some of the variables is the current Rector behaviour so we might aswell keep it like that. 

If this gets merged the next thing that might have to be updated is the doctrine ORM from v2 to V3 but doing that here makes it a bit too much in one PR.

MICMathieu commented:
@DennisdeBest some news about this PR ?

jgrygierek commented:
Anyone alive here?

jgrygierek commented:
Bump?

DennisdeBest commented:
@MICMathieu sorry no news so far

jgrygierek commented:
R.I.P DoctrineBehaviors?

DennisdeBest committed:
Update DoctrineBehaviors for Symfony 7 and Doctrine ORM 3 compatibility

This commit upgrades dependencies and adjusts classes to align with Symfony 7 and Doctrine ORM 3 requirements. Key changes include modifying type hints, updating deprecated `ClassMetadataInfo` usage to `ClassMetadata`, and refining mapping logics. This ensures compatibility with the latest frameworks and improves maintainability.

jgrygierek commented:
Bump? xD

acrobat commented:
I'm trying to move the package forward and add compatibility to newer versions of Symfony and Doctrine. See https://github.com/KnpLabs/DoctrineBehaviors/issues/743#issuecomment-3315270632 and #782. So please test these changes in your project and provide feedback if you have the time. Thanks!

acrobat commented:
Thanks for your work @DennisdeBest! This is now implemented through https://github.com/KnpLabs/DoctrineBehaviors/pull/782
